immer/detail/hamts/champ.hpp

```diff
@@ -129,11 +129,11 @@ struct champ
         for (auto i = count_t{}; i < max_depth<B>; ++i) {
             auto bit = bitmap_t{1u} << (hash & mask<B>);
             if (node->nodemap() & bit) {
-                auto offset = popcount(node->nodemap() & (bit - 1));
+                auto offset = popcount(static_cast<bitmap_t>(node->nodemap() & (bit - 1)));
```

Comment:
> I was fighting with integer propagation for some time and the only solution I found so far is to use casts for non-const expressions. 

In this quote, are you referring to this kind of cast? I agree that it looks a bit ugly and I don't fully understand why it is needed, both `bit` and `nodemap()` are `bitmap_t`. Could it maybe be alternatively solved by using `bitmat_t{1}` for the `1` literal at the end of the expression?

Reply:
I tried to make it like that: `bitmap_t{node->nodemap() & bitset_t{bit - bitset_t{1u}}}`
while making `bit` to be explicitly of type `const bitset_t` and making `node->nodemap()` return `bitset_t`, but it still propagates to int. Actually, as far as I understand, it does it for both minus and "bitwise and" operations.
I really tried my best to make it work without casts, but compiler wins this fight

Reply:
Hmmm, weird! Which compiler was exhibiting such behavior? I'd like to investigate this one before we merge.

Reply:
Both clang 3.9 and gcc 8.2

Reply:
Hi, sorry for long reply. I have refreshed the issue in my head and I don't really see any better solution.
In case `bitmap_t` is `uint16_t` aforementioned expression `node->nodemap() & (bit - 1)` yields `unsigned int` type and not `uint16_t`, hence following `popcount()` call will not be adequately resolved.

Probably, it would be possible to somehow tie `popcount` to a type.
E.g. making template `popcount()` with explicit template argument.
```
popcount<bitmap_t>(node->nodemap() & (bit - 1));
```

Anyways, `unsigned int` -> `bitmap_t` conversion would be required somewhere.

Reply:
As a follow up:
there might be additional `inline auto popcount_wrapper(bitmap_t)` declared after `bitmap_t` declaration, so it could isolate all those conversions.
P.S. function name is just a placeholder :)

Reply:
I like this latter approach! Do you want to do it?

Reply:
Yeah, I definitely want to bring this PR to finished state finally :)

Reply:
Super! Thank you so much!